internal/providers/msentraid/himmelblau_c.go

```diff
@@ -0,0 +1,330 @@
+		}
+	} else {
+		cCertKey = nil
+	}
+
+	msalErr := C.broker_init(
+		cAuthority,
+		cClientID,
+		cTransportKey,
+		cCertKey,
+		(**C.BrokerClientApplication)(unsafe.Pointer(&broker)),
+	)
+	if msalErr != nil {
+		return nil, fmt.Errorf("failed to initialize broker client: %v", C.GoString(msalErr.msg))
+	}
+
+	return broker, nil
+}
+
+func initEnrollAttrs(domain, hostname, osVersion string) (attrs *C.EnrollAttrs, err error) {
+	cDomain := C.CString(domain)
+	defer C.free(unsafe.Pointer(cDomain))
+	cHostname := C.CString(hostname)
+	defer C.free(unsafe.Pointer(cHostname))
+	cOSVersion := C.CString(osVersion)
+	defer C.free(unsafe.Pointer(cOSVersion))
+
+	msalErr := C.enroll_attrs_init(
+		cDomain,
+		cHostname,
+		nil, /* device_type - default is "Linux" */
+		0,   /* join_type - 0: Azure AD join */
+		cOSVersion,
+		&attrs,
+	)
+	if msalErr != nil {
+		return nil, fmt.Errorf("failed to initialize enroll attributes: %v", C.GoString(msalErr.msg))
+	}
+
+	// TODO: Do we not have to free the attrs?
+
+	return attrs, nil
+}
+
+func generateAuthValue() (authValue string, err error) {
+	var cAuthValue *C.char
+	if msalErr := C.auth_value_generate(&cAuthValue); msalErr != nil {
+		return "", fmt.Errorf("failed to generate auth value: %v", C.GoString(msalErr.msg))
+	}
+	defer C.free(unsafe.Pointer(cAuthValue))
+
+	return C.GoString(cAuthValue), nil
+}
+
+func createTPMMachineKey(tpm *boxedDynTPM, authValue string) (key *C.LoadableMachineKey, cleanup func(), err error) {
+	cAuthValue := C.CString(authValue)
+	defer C.free(unsafe.Pointer(cAuthValue))
+
+	var loadableMachineKey *C.LoadableMachineKey
+	msalErr := C.tpm_machine_key_create((*C.BoxedDynTpm)(unsafe.Pointer(tpm)), cAuthValue, &loadableMachineKey)
+	if msalErr != nil {
+		return nil, nil, fmt.Errorf("failed to create loadable machine key: %v", C.GoString(msalErr.msg))
+	}
+
+	cleanup = func() { C.loadable_machine_key_free(loadableMachineKey) }
+
+	return loadableMachineKey, cleanup, nil
+}
+
+func loadTPMMachineKey(tpm *boxedDynTPM, authValue string, loadableMachineKey *C.LoadableMachineKey) (key *C.MachineKey, cleanup func(), err error) {
+	cAuthValue := C.CString(authValue)
+	defer C.free(unsafe.Pointer(cAuthValue))
+
+	if msalErr := C.tpm_machine_key_load((*C.BoxedDynTpm)(unsafe.Pointer(tpm)), cAuthValue, loadableMachineKey, &key); msalErr != nil {
+		return nil, nil, fmt.Errorf("failed to load TPM machine key: %v", C.GoString(msalErr.msg))
+	}
+
+	cleanup = func() { C.machine_key_free(key) }
+
+	return key, cleanup, nil
+}
+
+func enrollDevice(broker *brokerClientApplication, refreshToken string, attrs *C.EnrollAttrs, tpm *boxedDynTPM, machineKey *C.MachineKey) (data *deviceRegistrationData, err error) {
+	cRefreshToken := C.CString(refreshToken)
+	defer C.free(unsafe.Pointer(cRefreshToken))
+
+	var cTransportKey *C.LoadableMsOapxbcRsaKey
+	var cCertKey *C.LoadableMsDeviceEnrolmentKey
+	var cDeviceID *C.char
```

Comment:
Needs to be free'd, `loadable_ms_oapxbc_rsa_key_free`, `loadable_ms_device_enrollment_key_free` and `free` I guess

Reply:
fixed